//
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto2";

import "zetasql/fuzzing/protobuf/parameter_grammar.proto";

package zetasql_expression_grammar;

message Expression {
    oneof expr_oneof {
        parameter_grammar.Value value = 1;
        CompoundExpr expr = 2;
    }
    required parameter_grammar.Default default_value = 3;
    required bool parenthesized = 4;
    optional parameter_grammar.Whitespace leading_pad = 5;
    optional parameter_grammar.Whitespace trailing_pad = 6;
}

message CompoundExpr {
    oneof compound_oneof {
        BinaryOperation binary_operation = 1;
    }
    required parameter_grammar.Default default_value = 2;
}

message BinaryOperation {
    enum Operator {
        PLUS = 0;
        MINUS = 1;
        MULTIPLY = 2;
        DIVIDE = 3;
    }
    required Operator op = 1;
    required Expression lhs = 2;
    required Expression rhs = 3;
    required parameter_grammar.Whitespace left_pad = 4;
    required parameter_grammar.Whitespace right_pad = 5;
}